### DISCUSSION TITLE
Nightmare scenario:  3 months worth of changes discarded -- any way to restore them?

### DISCUSSION THREAD

**Discussion by edmundweir** [community/community#144519] [Repositories] [answered]
### Select Topic Area

How to restore discarded changes from Github Desktop on Windows 10?

### Body

Hi there,

Like the title says, I hit a nightmare scenario today. I am developing a game and my repository had become too large to push my files, and so I had told myself I would update the .gitignore when I was finished with the present stage of work.

Today I finished the integration I was working on, and so I went to discard the commits I had made but not pushed as they were too large in file size, because I wanted to do a new commit with the updated .gitignore that should be within the file size limit. To my horror, when I discarded the changes in Github Desktop, Github went and removed every single change I had made over the last 3 months from the project files on my computer.

Is there any way to retrieve or restore these changes? They are not in the recycling bin. I am using Windows 10 and Github Desktop. I have not yet turned off the machine, though I did close Visual Studio Code.

Any help would obviously be immensely appreciated! Thank you!!

**Comment by versenilvis** [comment, marked as answer]
### 1. **Check GitHub Desktop's History and Reflog**
   GitHub Desktop itself doesn't keep backups of discarded changes, but the underlying Git system may have a way to restore them via the reflog (a record of changes in Git). Here's how to try to recover them:

   - **Open Git Bash**: You’ll need to use Git Bash for this recovery process.
   - **Check the Reflog**: Run the following command in your project directory (where the repository is located):
     ```bash
     git reflog
     ```
     The reflog keeps a record of all recent changes, even those that are not committed, and should include entries for the changes you discarded. You should look for any recent commits or changes that were discarded.

   - **Restore the Changes**: If you find the commit or state you're looking for in the reflog, you can check it out or reset to it. For example, if you want to restore to a specific commit in the reflog:
     ```bash
     git checkout <reflog_entry>
     ```
     Or if you want to restore the changes to your working directory:
     ```bash
     git reset --hard <reflog_entry>
     ```

### 2. **Check for Local Backups**
   Sometimes, certain IDEs (like Visual Studio Code) create local backups or caches of files that can be restored.

   - **Look in the VS Code Backups Folder**: Open your project folder in File Explorer and search for any VS Code backups or temporary files (you might find them in folders like `.vscode`, or in your project's directory, depending on the settings).
   - **Look for Auto-Saved Files**: Some IDEs have automatic save or backup features. If you have these enabled in VS Code, you may be able to recover a previous version of the file.

### 3. **Look for a Stash (If You Stashed Changes)**
   If you used Git's stash feature in the past, you may have saved your changes without realizing it. To check for stashes, run:
   ```bash
   git stash list
   ```
   If there’s any stash listed, you can restore it using:
   ```bash
   git stash apply
   ```

### 4. **Data Recovery Software**
   If the files are truly gone and not in the Git history or stash, you might still be able to use **file recovery software**. These tools scan your hard drive for deleted files that haven’t been overwritten yet. Some popular options are:
   - **Recuva** (a free tool that can recover deleted files)
   - **EaseUS Data Recovery Wizard**
   
   These tools might be able to recover lost files, but it's important to act quickly before new data overwrites the deleted content.

### 5. **Preventing This in the Future**
   Moving forward, consider using these best practices to avoid similar situations:
   - **Commit Regularly**: Even if your repository gets too large, committing often ensures that changes are safely stored in Git.
   - **Use `.gitignore` Early**: It’s better to update `.gitignore` sooner rather than later, and always commit the `.gitignore` file to keep your repository clean.
   - **Create Backups**: Consider setting up automated backups or using cloud storage services (like OneDrive, Dropbox) to keep important files safe.

If this is helpful just help me to mark this as answer

**Reply by edmundweir** [reply]
Thank you so much, kind human!

`git reset --hard <reflog_entry>` did the job. :)

**Reply by queenofcorgis** [reply]
So glad you were able to get this solved @edmundweir ❤️ thanks for posting in the community!
